vm/src/hint_processor/builtin_hint_processor/bootloader/fact_topologies.rs

```diff
@@ -0,0 +1,346 @@
+use crate::hint_processor::builtin_hint_processor::bootloader::types::PackedOutput;
+use crate::types::errors::math_errors::MathError;
+use crate::types::relocatable::Relocatable;
+use crate::vm::errors::hint_errors::HintError;
+use crate::vm::errors::runner_errors::RunnerError;
+use crate::vm::runners::builtin_runner::OutputBuiltinRunner;
+use felt::Felt252;
+use serde::Serialize;
+use std::fs::File;
+use std::path::Path;
+
+#[derive(Clone, Debug, PartialEq, Serialize)]
+pub struct FactTopology {
+    #[allow(dead_code)]
+    pub tree_structure: Vec<Felt252>,
+    pub page_sizes: Vec<usize>,
+}
+
+#[derive(Serialize)]
+struct FactTopologyFile<'a> {
+    fact_topologies: Vec<&'a FactTopology>,
+}
+
+impl AsRef<FactTopology> for FactTopology {
+    fn as_ref(&self) -> &FactTopology {
+        self
+    }
+}
+
+#[derive(Debug, thiserror_no_std::Error)]
+pub enum FactTopologyError {
+    #[error("Expected {0} fact topologies but got {1}")]
+    WrongNumberOfFactTopologies(usize, usize),
+
+    #[error("Composite packed outputs are not supported yet")]
+    CompositePackedOutputNotSupported(PackedOutput),
+
+    #[error("Could not add page to output: {0}")]
+    FailedToAddOutputPage(#[from] RunnerError),
+
+    #[error(transparent)]
+    Math(#[from] MathError),
+}
+
+impl Into<HintError> for FactTopologyError {
+    fn into(self) -> HintError {
+        match self {
+            FactTopologyError::WrongNumberOfFactTopologies(_, _) => {
+                HintError::AssertionFailed(self.to_string().into_boxed_str())
+            }
+            FactTopologyError::CompositePackedOutputNotSupported(_) => HintError::CustomHint(
+                "Cannot compute fact topologies for composite packed outputs (yet)"
+                    .to_string()
+                    .into_boxed_str(),
+            ),
+            FactTopologyError::FailedToAddOutputPage(e) => HintError::CustomHint(
+                format!("Cannot add output page to output segment: {}", e).into_boxed_str(),
+            ),
+            FactTopologyError::Math(e) => HintError::Math(e),
+        }
+    }
+}
+
+#[derive(thiserror_no_std::Error, Debug)]
+pub enum WriteFactTopologiesError {
+    #[error("Failed to create fact topology file: {0}")]
+    Io(#[from] std::io::Error),
+    #[error("Failed to serialize fact topologies: {0}")]
+    Serialization(#[from] serde_json::Error),
+}
+
+impl Into<HintError> for WriteFactTopologiesError {
+    fn into(self) -> HintError {
+        HintError::CustomHint(self.to_string().into_boxed_str())
+    }
+}
+
+/// Flattens and extracts the fact topologies from packed outputs.
+///
+/// Note that `packed_outputs` and `fact_topologies` must have the same length.
+///
+/// * `packed_outputs`: Packed outputs.
+/// * `fact_topologies`: Fact topologies.
+pub fn compute_fact_topologies<'a>(
+    packed_outputs: &Vec<PackedOutput>,
+    fact_topologies: &'a Vec<FactTopology>,
+) -> Result<Vec<&'a FactTopology>, FactTopologyError> {
```

Comment:
is the `'a` lifetime here to help avoid copying FactTopologies? I'm guessing their vecs are expected to get quite large?

Reply:
No idea how large it gets, but I thought using references here wouldn't hurt.